rllib/algorithms/algorithm_config.py

```diff
@@ -638,8 +641,26 @@ def update_from_dict(
                 if isinstance(value, str):
                     value = deserialize_type(value, error=True)
                 self.callbacks(callbacks_class=value)
+            elif key == "action_space":
+                self.action_space = gym_space_from_dict(value)
             elif key == "env_config":
+                # Some values in env_config might be encoded space dicts. We should
+                # deserialize these here.
+                def deserialize_env_config(s):
+                    if isinstance(s, dict) and "space" in s:
```

Comment:
Actually, no, `__struct` is the struct of the entire `env_config` dict. It could have non-Space-encoding dicts nested inside, which then would NOT have the "space" key in them. So it's important to check for this. We only need `__struct` because we wouldn't want to tree_map_structure into the encoded Space-dicts, but only up to them.